### ISSUE TITLE
int(math.Inf(1)) in funcOpMul is incorrect

### ISSUE THREAD

**Issue by Ikke** [itchyny/gojq#69]
This was caught out by a test failure on mips64:

```sh
--- FAIL: TestCliRun (0.86s)
    --- FAIL: TestCliRun/multiply_strings (0.00s)
        cli_test.go:87: standard output:
              (
              	"""
              	... // 8 identical lines
              	"abcabc"
              	null
            - 	null
            + 	""
              	"""
              )
            
FAIL
FAIL	github.com/itchyny/gojq/cli	0.872s
```

The result of `int(math.Inf(1))`<sup>[1][]</sup> depends on the architecture, and is not correct in general to do. 

On s390x / ppc64le (big-endian), it returns 9223372036854775807, which would have failed as well, would it not be for the limit condition.

A simple way to fix this is to check:

```go
math.IsNaN(cnt - cnt)
```

Which catches both infinite (positive / negative) as well as `NaN`.

[1]:https://github.com/itchyny/gojq/blob/main/operator.go#L381

**Comment by itchyny** [comment]
`gojq -n '[1] | has(infinite)'` fails to print `false` on mips64?

**Comment by itchyny** [comment]
Or more simply `[1][infinite]`? I hope the sign of `int(math.Inf(1))` does not make the tests fail but the mips64 result is just surprising. It breaks my various assumptions. gojq cannot support mips64 because I don't have CI for it, so even if I include your patch, the support is best-effort and may break in the future.

**Comment by itchyny** [comment]
Another possible test case: `1 % infinite`

**Comment by itchyny** [comment]
I added suspicious test cases to the main branch, also fixed array indexing and slicing. I noticed that `[1][:1e300]` yields `[]` incorrectly with gojq 0.12.2 so the found bug appears to spread to the architectures gojq officially supports. Thanks.

**Comment by Ikke** [comment]
> `gojq -n '[1] | has(infinite)'` fails to print `false` on mips64?

Correct

> Or more simply `[1][infinite]`? I hope the sign of `int(math.Inf(1))` does not make the tests fail but the mips64 result is just surprising. It breaks my various assumptions. gojq cannot support mips64 because I don't have CI for it, so even if I include your patch, the support is best-effort and may break in the future.

mips64 is best-effort for us as well, so we appreciate any effort, but do not expect nothing to break. We do always run the test suite to catch gotcha's like these.

Our mips64 is softfloat, as the hardware FPU is buggy. And the software fpu is probably more strict in this regard.

**Comment by itchyny** [comment]
@Ikke If you still have the mips64 environment, can you test the main branch HEAD please?

**Comment by Ikke** [comment]
> @Ikke If you still have the mips64 environment, can you test the main branch HEAD please?

I now get new test failures:

```
--- FAIL: TestCliRun (0.88s)
    --- FAIL: TestCliRun/infinite_function (0.00s)
        cli_test.go:84: exit code: got: 5, expected: 0
        cli_test.go:87: standard output:
              (
                """
                ... // 16 identical lines
                [null,false,false,false]
                [null,false,false,false]
            -   [0,false,false,false]
            -   [0,false,false,false]
            -   [0,false,false,false]
            -   [0,false,false,false]
            -   [0,false,false,false]
            -   [-0,false,false,false]
            -   [-0,false,false,false]
            -   [0,false,false,false]
            -   [1.7976931348623157e+308,true,false,true]
            -   [-1.7976931348623157e+308,false,true,false]
            -   [1.7976931348623157e+308,true,false,true]
            -   [-1.7976931348623157e+308,false,true,false]
                """
              )

        cli_test.go:90: standard error output:
              string(
            -   "",
            +   "gojq: cannot modulo number (1.7976931348623157e+308) by: number (1.7976931348623157e+308)\n",
              )
```

**Comment by itchyny** [comment]
Thanks, I expected that failure to happen and was wondering how to handle this. `jq -n '1%infinite'` would fail to print 1 too.

**Comment by itchyny** [comment]
I fixed the modulo operator.

**Comment by Ikke** [comment]
Thanks, I can confirm all tests pass again with the latest commit.

**Comment by mpolanski** [comment]
Looks like gojq 0.12.3 introduced a regression that made the `infinite_function` test failing again. x86 and x86_64 are fine, but aarch64, armv7, ppc64le and s390x (not sure about mips64) have this error:
```
--- FAIL: TestCliRun (0.29s)
    --- FAIL: TestCliRun/infinite_function (0.00s)
        cli_test.go:87: standard output:
              (
              	"""
              	... // 17 identical lines
              	[null,false,false,false]
              	[0,false,false,false]
            - 	[0,false,false,false]
            - 	[0,false,false,false]
            + 	[-1,false,false,false]
            + 	[9223372036854775807,false,false,false]
              	[0,false,false,false]
              	[0,false,false,false]
              	... // 8 identical lines
              	"""
              )
            
FAIL
FAIL	github.com/itchyny/gojq/cli	0.299s
```

**Comment by itchyny** [comment]
Okay, so on some architectures `int(-inf) != int(inf)`. It seems that `jq -n '[infinite, -infinite] | .[]%.[]'` (and not limited to infinities; `jq -n '[1e30, -1e30] | .[]%.[]'`) yields different result on x86_64 and aarch64.

**Comment by itchyny** [comment]
BTW thit is not a regression, the cli yields the same result but added tests included the different behavior on architectures. I'll fix this anyway.

**Comment by itchyny** [comment]
Fixed by e2d8db6b19f2ef6f4b18ef21dc830b4dad988e8d.

**Comment by mpolanski** [comment]
The test passes now. Thanks!
